File: webapi/Models/Flexibility/Efficiency/SelfExplanation.cs
using webapi.Models.User;

namespace webapi.Models.Flexibility
{
    public class SelfExplanation
    {
        public Method Method { get; set; }

        public bool IsSingleChoice { get; set; }

        public List<Option> Options { get; set; }

        public SelfExplanation() { }

        public SelfExplanation(ExtendedSelfExplanation explanation, Language language)
        {
            Method = explanation.Method;
            IsSingleChoice = explanation.IsSingleChoice;
            Options = explanation.Options.Select(option => new Option(option, language)).ToList();
        }
    }

    public class Option
    {
        public string Text { get; set; }

        public bool IsSolution { get; set; }

        public string? Reason { get; set; }

        public Option() { }

        public Option(ExtendedOption extendedOption, Language language)
        {
            Text = language == Language.de ? extendedOption.TextDE : extendedOption.TextEN;
            IsSolution = extendedOption.IsSolution;
            Reason = language == Language.de ? extendedOption.ReasonDE : extendedOption.ReasonEN;
        }
    }
}

File: reactapp/src/types/flexibility/selfExplanation.ts
import { Method } from "@/types/flexibility/enums.ts";
import { Type } from "class-transformer";

export interface IOption {
    readonly text: string;
    readonly isSolution: boolean;
    readonly reason?: string | undefined | null;
}

export class Option implements IOption {
    public readonly text: string;

    public readonly isSolution: boolean;

    public readonly reason?: string | null;

    constructor(text: string, isSolution: boolean, reason?: string | null) {
        this.text = text;
        this.isSolution = isSolution;
        this.reason = reason;
    }
}

export interface ISelfExplanation {
    readonly method: Method;
    readonly isSingleChoice: boolean;
    readonly options: IOption[];
}

export class SelfExplanation implements ISelfExplanation {
    public readonly method: Method;

    public readonly isSingleChoice: boolean;

    @Type(() => Option)
    public options: Option[];

    constructor(method: Method, isSingleChoice: boolean, options: Option[]) {
        this.method = method;
        this.isSingleChoice = isSingleChoice;
        this.options = options;
    }
}